axum-extra/src/multipart_builder.rs

```diff
@@ -0,0 +1,337 @@
+        for part in self.parts {
+            // for each part, the boundary is preceded by two dashes
+            serialized_form.extend_from_slice(format!("--{}\r\n", boundary).as_bytes());
+            serialized_form.extend_from_slice(&part.serialize());
+        }
+        serialized_form.extend_from_slice(format!("--{}--", boundary).as_bytes());
+        (headers, serialized_form).into_response()
+    }
+}
+
+impl Default for MultipartForm {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+// Every part is expected to contain:
+// - a [Content-Disposition](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition
+// header, where `Content-Disposition` is set to `form-data`, with a parameter of `name` that is set to
+// the name of the field in the form. In the below example, the name of the field is `user`:
+// ```
+// Content-Disposition: form-data; name="user"
+// ```
+// If the field contains a file, then the `filename` parameter may be set to the name of the file.
+// Handling for non-ascii field names is not done here, support for non-ascii characters may be encoded using
+// methodology described in RFC 2047.
+// - (optionally) a `Content-Type` header, which if not set, defaults to `text/plain`.
+// If the field contains a file, then the file should be identified with that file's MIME type (eg: `image/gif`).
+// If the `MIME` type is not known or specified, then the MIME type should be set to `application/octet-stream`.
+// - If the part does not conform to the default encoding, then the `Content-Transfer-Encoding` header may be supplied.
+// Valid settings for that header are: "base64", "quoted-printable", "8bit", "7bit", and "binary".
+/// A single part of a multipart form as defined by
+/// <https://www.w3.org/TR/html401/interact/forms.html#h-17.13.4>
+/// and RFC2388.
+#[derive(Debug)]
+pub struct Part {
+    /// The name of the part in question
+    name: String,
+    /// If the part should be treated as a file, the filename that should be attached that part
+    filename: Option<String>,
+    /// The `Content-Type` header. While not strictly required, it is always set here
+    mime_type: String,
+    /// The content/body of the part
+    contents: Vec<u8>,
+    /// The encoding that the contents should be encoded under
+    encoding: TransferEncoding,
+}
+
+impl Part {
+    /// Create a new part with `Content-Type` of `text/plain` with the supplied name and contents.
+    /// This form will not have a defined file name.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use axum_extra::multipart_builder::{MultipartForm, Part};
+    ///
+    /// // create a form with a single part that has a field with a name of "foo",
+    /// // and a value of "abc"
+    /// let parts: Vec<Part> = vec![Part::text("foo", "abc")];
+    /// let form = MultipartForm::with_parts(parts);
+    /// ```
+    pub fn text(name: &str, contents: &str) -> Self {
```

Comment:
Generally, it makes sense to require `String` if that's what you'll need.

Currently, if someone had a large string content, instead of moving it into this constructor (assuming the caller doesn't need it anymore), they would pass only slice here and this method would clone it for no reason.